### PR TITLE
[CUDA] Add an optimized channels-last path for InstanceNorm

### PR DESCRIPTION
**Pull request by xwang233**
This PR adds channels-last paths for **InstanceNorm** for `{cpu, gpu} x InstanceNorm{2, 3}d`.

For a tensor with size NHWC on CUDA, an optimized path by using cudnn batch_norm as the backend of InstanceNorm is enabled. The implementation splits the tensor in the batch dimension, and **call cudnn batch_norm N times with a 1HWC tensor** each. Then, the normalized tensors are "stacked" in the batch dimension to give the output.

The current batch_norm implementation on master branch allocates memory spaces for output tensors in the function, which is inconvenient for this InstanceNorm implementation. This PR allow batch_norm to take pre-allocated memory spaces as output. With that, extra copies for the output tensors are avoided and performances are largely improved in our internal models.

For all other tensor backends or tensors that can't be optimized by cudnn batch_norm, the current implementation of InstanceNorm will still be used. We just transform the output to the same memory format as the input at the end.

**BC breaking changes**: For functions below, we add some extra input arguments as pre-allocated output memory spaces. If those arguments are given as `c10::nullopt` to the functions, nothing changed. The batch_norm functions will still allocate output memory spaces as before.

- `at::cudnn_batch_norm_backward`
- `at::cudnn_batch_norm`
- `at::_batch_norm_impl_index_backward`
- `at::_batch_norm_impl_index`

### CODE REVIEW

**Comment by jjsjann123** [comment]
Oops. I think I messed up big time with the rebase.. :cry: Let me try again.

**Comment by jjsjann123** [comment]
errrr. Sorry about the requested reviews, I think that came from my previous push... No idea what went wrong there.. :confused:

**Comment by xwang233** [comment]
Hi reviewers, this PR is ready for review. cc @ezyang @albanD @soulitzer

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -844,9 +844,9 @@
   dispatch:
     QuantizedCPU: quantized_batch_norm
 
-- func: _batch_norm_impl_index(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled) -> (Tensor, Tensor, Tensor, Tensor, int)
+- func: _batch_norm_impl_index(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled, Tensor? output_inplace, Tensor? save_mean_inplace, Tensor? save_var_inplace) -> (Tensor, Tensor, Tensor, Tensor, int)
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:847]: The alias annotations here don't look right, the new inplace arguments are mutable amirite?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -2386,6 +2386,14 @@
 - func: instance_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool use_input_stats, float momentum, float eps, bool cudnn_enabled) -> Tensor
   variants: function
 
+- func: _instance_norm_channels_last(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool use_input_stats, float momentum, float epsilon, bool cudnn_enabled) -> (Tensor, Tensor, Tensor, Tensor, int)
+  variants: function
+  dispatch:
+    CPU, CUDA: _instance_norm_channels_last
+
+- func: _instance_norm_channels_last_backward(int impl_index, Tensor input, Tensor grad, Tensor? weight_opt, Tensor? running_mean_opt, Tensor? running_var_opt, Tensor? save_mean_opt, Tensor? save_var_opt, bool use_input_stas, float epsilon, Tensor reservedSpace) -> (Tensor, Tensor, Tensor)
+  variants: function
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:2395]: It's not clear to me why you need `native_functions.yaml` entries for these

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1438,12 +1438,12 @@
...
 
-- func: cudnn_batch_norm(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon) -> (Tensor, Tensor, Tensor, Tensor)
+- func: cudnn_batch_norm(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon, Tensor? output_inplace, Tensor? save_mean_inplace, Tensor? save_var_inplace) -> (Tensor, Tensor, Tensor, Tensor)
   dispatch:
     CUDA: cudnn_batch_norm
 
 # NB: You can only use this if you used cudnn_batch_norm training=True
-- func: cudnn_batch_norm_backward(Tensor input, Tensor grad_output, Tensor weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_var, float epsilon, Tensor reserveSpace) -> (Tensor, Tensor, Tensor)
+- func: cudnn_batch_norm_backward(Tensor input, Tensor grad_output, Tensor weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_var, float epsilon, Tensor reserveSpace, Tensor? grad_input_inplace, Tensor? grad_weight_inplace, Tensor? grad_bias_inplace) -> (Tensor, Tensor, Tensor)
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:1446]: @jbschlosser were you planning to do any batch norm consolidation in the future haha 📦

**Comment by ngimel** [comment]
This PR is stalled pending the dedicated kernel for instance norm (or nvfuser-generated kernel for instance norm), calling cudnn batch norm in a loop is not a good approach.
